test/unit-tests/marginCalculatorNakedMargin.test.ts

```diff
@@ -0,0 +1,848 @@
+import {
+  MockERC20Instance,
+  CalculatorTesterInstance,
+  MockAddressBookInstance,
+  MockOracleInstance,
+  MockOtokenInstance,
+} from '../../build/types/truffle-types'
+import {
+  createScaledNumber as scaleNum,
+  createScaledBigNumber as scaleBigNum,
+  createVault,
+  createScaledBigNumber,
+  createScaledNumber,
+} from '../utils'
+import {assert} from 'chai'
+import BigNumber from 'bignumber.js'
+
+const {expectRevert, time} = require('@openzeppelin/test-helpers')
+const MockAddressBook = artifacts.require('MockAddressBook.sol')
+const MockOracle = artifacts.require('MockOracle.sol')
+const MockOtoken = artifacts.require('MockOtoken.sol')
+const MockERC20 = artifacts.require('MockERC20.sol')
+const MarginCalculator = artifacts.require('CalculatorTester.sol')
+const ZERO_ADDR = '0x0000000000000000000000000000000000000000'
+
+const expectedRequiredMargin = (
+  shortAmount: number,
+  strikePrice: number,
+  underlyingPrice: number,
+  isPut: boolean,
+  upperBoundValue: number,
+  spotShockValue: number,
+) => {
+  let a, b, marginRequired
+  // if (isPut) {
+  //   a = BigNumber.minimum(strikePrice, spotShockValue.times(underlyingPrice).dividedBy(1e27))
+  //   b = BigNumber.maximum(strikePrice.minus(spotShockValue.times(underlyingPrice).dividedBy(1e27)), new BigNumber(0))
+  //   marginRequired = upperBoundValue.times(a).plus(b).times(shortAmount).dividedBy(1e27)
+  // } else {
+  //   const one = scaleBigNum(1, 27)
+  //   a = BigNumber.minimum(one, strikePrice.dividedBy(underlyingPrice.dividedBy(spotShockValue)))
+  //   b = BigNumber.maximum(one.minus(strikePrice.dividedBy(underlyingPrice.dividedBy(spotShockValue))), new BigNumber(0))
+  //   marginRequired = upperBoundValue.times(a).plus(b).times(shortAmount);
+  // }
+
+  if (isPut) {
+    a = Math.min(strikePrice, spotShockValue * underlyingPrice)
+    b = Math.max(strikePrice - spotShockValue * underlyingPrice, 0)
+    marginRequired = (upperBoundValue * a + b) * shortAmount
+  } else {
+    a = Math.min(1, strikePrice / (underlyingPrice / spotShockValue))
+    b = Math.max(1 - strikePrice / (underlyingPrice / spotShockValue), 0)
+    marginRequired = (upperBoundValue * a + b) * shortAmount
+  }
+
+  return marginRequired
+}
+
+const calcRelativeDiff = (expected: BigNumber, actual: BigNumber): BigNumber => {
+  let diff: BigNumber
+
+  if (actual.isGreaterThan(expected)) {
+    diff = actual.minus(expected)
+  } else {
+    diff = expected.minus(actual)
+  }
+  return diff
+}
+
+contract('MarginCalculator', ([owner, random]) => {
+  let expiry: number
+
+  let calculator: CalculatorTesterInstance
+  let addressBook: MockAddressBookInstance
+  let oracle: MockOracleInstance
+
+  let usdc: MockERC20Instance
+  let dai: MockERC20Instance
+  let weth: MockERC20Instance
+
+  const usdcDecimals = 6
+  const daiDecimals = 8
+  const wethDecimals = 18
+
+  const vaultType = 1
+
+  const errorDelta = 0.1
```

Comment:
where is the error delta coming from? Is it js or solidity?

Reply:
there is a rounding issue in the get margin function for the naked vault case, so I explicitly rounded up for the getMargin, and making sure the rounding issue is not bigger than 0.1 for a position of 1 size. (it is a solidity issue)

Reply:
Planning to explicitly round up for returning the margin required and round down for the liquidation price.